language/move-prover/bytecode/src/spec_translator.rs

```diff
@@ -112,29 +120,64 @@ impl TranslatedSpec {
         !self.aborts_with.is_empty() || self.aborts.iter().any(|(_, _, c)| c.is_some())
     }
 
-    /// Return an iterator of effective pre conditions. Each individual condition is or-ed with
-    /// the aborts condition, if present, as the pre-condition only needs to hold if the function
-    /// does not abort.
-    pub fn pre_conditions(&self, builder: &FunctionDataBuilder<'_>) -> Vec<(Loc, Exp)> {
-        let abort_cond = self.aborts_condition(builder);
-        self.pre
-            .iter()
-            .map(|(loc, exp)| {
-                (
-                    loc.clone(),
-                    builder
-                        .mk_join_opt_bool(ast::Operation::Or, Some(exp.clone()), abort_cond.clone())
-                        .unwrap(),
-                )
-            })
-            .collect()
+    /// Return an iterator of effective pre conditions.
+    pub fn pre_conditions(
+        &self,
+        _builder: &FunctionDataBuilder<'_>,
+    ) -> impl Iterator<Item = (Loc, Exp)> + '_ {
+        self.pre.iter().cloned()
+    }
+
+    /// Returns a sequence of EventCheck expressions which verify the `emits` clauses of a
+    /// function spec. While logically we could generate a single EventCheck, for better
+    /// error reporting we construct incrementally multiple EventCheck expressions with some
+    /// redundancy for each individual `emits, so we the see the exact failure at the right
+    /// emit condition.
+    pub fn emits_conditions(&self, builder: &FunctionDataBuilder<'_>) -> Vec<(Loc, Exp)> {
```

Comment:
@junkil-park Here is where we now generate the condition for checking the emit conditions. We can express this by a combination of AST operators `EmptyEventStore, `ExtendEventStore`, and `CheckEventStore`.